HW3/P5/label_regions.cl

```diff
@@ -80,18 +94,50 @@ propagate_labels(__global __read_write int *labels,
     old_label = buffer[buf_y * buf_w + buf_x];
 
     // CODE FOR PARTS 2 and 4 HERE (part 4 will replace part 2)
-    
+    // Part 2, update buffer for those within the threshold only.
+    /*
+    if (old_label < w*h) {
+        buffer[buf_y * buf_w + buf_x] = labels[old_label];
+    }
+    */
+
+    if (idx_1D == 0) {
+        int last_fetch = -1;
+        int last_idx = -1;
+        for (int yiter = halo; yiter < buf_h - halo; ++yiter) {
+            for (int xiter = halo; xiter < buf_w - halo; ++xiter) {
+                int cur_idx = yiter * buf_h + xiter;
+                if (buffer[cur_idx] >= w*h) continue;
+                if (buffer[cur_idx] != last_idx) {
+                    last_idx = cur_idx;
+                    last_fetch = labels[buffer[cur_idx]];
+                }
+                buffer[cur_idx] = last_fetch;
+            }
+        }
+    }
+
+    barrier(CLK_LOCAL_MEM_FENCE);
+
     // stay in bounds
     if ((x < w) && (y < h)) {
         // CODE FOR PART 1 HERE
         // We set new_label to the value of old_label, but you will need
         // to adjust this for correctness.
-        new_label = old_label;
+        if (old_label >= w * h) new_label = old_label;
+        else new_label = update_label(
+                            old_label,
```

Comment:
After parts 2 and 4, you should use buffer[buf_w * buf_y + buf_x] instead of old_label.